homeassistant/helpers/entity.py

```diff
@@ -1361,6 +1367,8 @@ async def __async_remove_impl(self, force_remove: bool) -> None:
             not force_remove
             and self.registry_entry
             and not self.registry_entry.disabled
+            # Check if entity is still in the entity registry
+            and not self._removed_from_registry
```

Comment:
There was a race here because `_async_registry_updated` would previously run in a task and `self.registry_entry` would not get unset soon enough so it would write unavailable state instead of removing the entity.  I've seen this problem happen in production but I never could figure out why until now.